connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedKafkaCluster.java

```diff
@@ -0,0 +1,314 @@
+        producerProps.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers());
+        producerProps.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArraySerializer");
+        producerProps.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArraySerializer");
+        producer = new KafkaProducer<>(producerProps);
+    }
+
+    private void stop() {
+
+        try {
+            producer.close();
+        } catch (Exception e) {
+            log.error("Could not shutdown producer ", e);
+        }
+
+        for (KafkaServer broker : brokers) {
+            try {
+                broker.shutdown();
+            } catch (Throwable t) {
+                log.error("Could not shutdown broker at {}", address(broker), t);
+            }
+        }
+
+        for (KafkaServer broker : brokers) {
+            try {
+                log.info("Cleaning up kafka log dirs at {}", broker.config().logDirs());
+                CoreUtils.delete(broker.config().logDirs());
+            } catch (Throwable t) {
+                log.error("Could not clean up log dirs for broker at {}", address(broker), t);
+            }
+        }
+
+        try {
+            zookeeper.shutdown();
+        } catch (Throwable t) {
+            log.error("Could not shutdown zookeeper at {}", zKConnectString(), t);
+        }
+    }
+
+    private static void putIfAbsent(final Properties props, final String propertyKey, final Object propertyValue) {
+        if (!props.containsKey(propertyKey)) {
+            props.put(propertyKey, propertyValue);
+        }
+    }
+
+    private String createLogDir() throws IOException {
+        TemporaryFolder tmpFolder = new TemporaryFolder();
+        tmpFolder.create();
+        return tmpFolder.newFolder().getAbsolutePath();
+    }
+
+    public String bootstrapServers() {
+        return address(brokers[0]);
+    }
+
+    public String address(KafkaServer server) {
+        return server.config().hostName() + ":" + server.boundPort(listenerName);
+    }
+
+    public String zKConnectString() {
+        return "127.0.0.1:" + zookeeper.port();
+    }
+
+    /**
+     * Create a Kafka topic with 1 partition and a replication factor of 1.
+     *
+     * @param topic The name of the topic.
+     */
+    public void createTopic(String topic) {
+        createTopic(topic, 1, 1, new HashMap<>());
+    }
+
+    /**
+     * Create a Kafka topic with the given parameters.
+     *
+     * @param topic       The name of the topic.
+     * @param partitions  The number of partitions for this topic.
+     * @param replication The replication factor for (partitions of) this topic.
+     * @param topicConfig Additional topic-level configuration settings.
+     */
+    public void createTopic(String topic, int partitions, int replication, Map<String, String> topicConfig) {
+        replication = Math.min(replication, brokers.length);
```

Comment:
we should just throw an exception if there aren't enough brokers. i think the behavior if we do this will just be confusing (something we've seen in practice too with the consumer offsets topic when there aren't enough active brokers and it gets created with too low of a replication factor).